### PR TITLE
Role and Everyone Mentions

### PR DESCRIPTION
**Pull request by NinekoTheCat**

### CODE REVIEW

**Comment by AlbertMarashi** [comment]
Looks good to me in terms of usage

**Comment by Zomatree** [comment]
The reference level explanation should go into detail on the actual implementation of it, including changes to routes, any new routes and changes to how the events will be sent to users.

It should have enough detail to be able to implement the feature with only the RFC.

**Comment by NinekoTheCat** [comment]
> The reference level explanation should go into detail on the actual implementation of it, including changes to routes, any new routes and changes to how the events will be sent to users.
> 
> It should have enough detail to be able to implement the feature with only the RFC.

@Zomatree not necessarily. The RFC is for a feature and the implementation is up to the developer. the reference level explanation goes over how it'd look on the bridge between backend and frontend but that's due to the fact they need to work together. any implementation on the backend or frontend side only needs to guarantee that a mention will be treated as a mention

**Comment by Zomatree** [comment]
The RFC must go into detail on how it will be implemented.

**Comment by NinekoTheCat** [comment]
> The RFC must go into detail on how it will be implemented.

@Zomatree 
I might as-well implement it if I go into that much detail.

**Comment by insertish** [comment]
It's much harder to discuss already written code, the RFC should present and discuss high level implementation and behaviour.
For example, in the case of mentions, how should events be fanned out? What are the advantages and disadvantages of different ways we could approach this, such as managing a queue in database, managing a queue in-memory, how the algorithm scale, etc.

**Comment by NinekoTheCat** [comment]
> It's much harder to discuss already written code, the RFC should present and discuss high level implementation and behaviour. For example, in the case of mentions, how should events be fanned out? What are the advantages and disadvantages of different ways we could approach this, such as managing a queue in database, managing a queue in-memory, how the algorithm scale, etc.

@insertish I'm sorry but this is my first RFC however I do discuss somewhat different solutions in the RFC itself? I do not quite know what you mean, sorry.

**Comment by insertish** [comment]
I'm specifically talking about how the server needs to handle bulk mentions, part of adding them is figuring out how the server can send push notifications efficiently.

Permissions values should also be proposed since this is relevant to everyone that implements the RFC.

**Comment by NinekoTheCat** [comment]
I'll edit the RFC and mention that thank you

**Comment by NinekoTheCat** [comment]
I edited the RFC to mention the potential implementation

**Review by nulldg — CHANGES_REQUESTED** [review]
i don't like opt-out `@everyone` pings, but i do support role pings. i think role pings are sufficient to satisfy both functions, particularly if roles can be self-assigned.

File: rfcs/rfc-0004-role-and-everyone-mentions.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+# Drawbacks
+
+- annoying users with notifications they did not want to be apart of
+- being abused by power-moderators
+- would add parsing time as it'd require the back-end to parse it further and do a complex selection over the affected users.
+# Rationale and alternatives
+
+- This solutions works the best as it's simple and we already have frontend components for mentions, and back-end infrastructure delivering notifications dynamically. It only requires additional parsing of content and easy to add permissions to happen. 
+- Adding multiple mentions by selecting affected roles or users using a table of some kind, this approach could work however there would need to be large changes done to the messaging API to facilitate such a solution.
+- A bot that sends a couple messages mentioning all of the affected usernames, I did not consider this to be a good solution as it practically DDoSes our servers and if used frequently could fill up chats with useless messages.
```
[the remark this answers is not in the document]
Comment (NinekoTheCat) [reply rfcs/rfc-0004-role-and-everyone-mentions.md:78]: what do you mean?

File: rfcs/rfc-0004-role-and-everyone-mentions.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+# Prior art
+
+Discord, Slack and telegram solve this issue similarly and it's worked for them so far.
```
[the remark this answers is not in the document]
Comment (NinekoTheCat) [reply rfcs/rfc-0004-role-and-everyone-mentions.md:82]: that's a valid concern a potential alteration I could do is remove @everyone altogether and just stick to role mentions

File: rfcs/rfc-0004-role-and-everyone-mentions.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (NinekoTheCat) [reply rfcs/rfc-0004-role-and-everyone-mentions.md:78]: this RFC only sticks to roles and everyone mentions, if you'd like for there to be a subscription system make a separate rfc

File: rfcs/rfc-0004-role-and-everyone-mentions.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (NinekoTheCat) [reply rfcs/rfc-0004-role-and-everyone-mentions.md:83]: although only you felt that way about them

[the remark this answers is not in the document]
Comment (JackDotJS) [reply rfcs/rfc-0004-role-and-everyone-mentions.md:83]: i sorta half disagree.

i believe `@everyone` pings can be quite useful for very small servers, especially servers made for groups of friends or coworkers and such. they can also be useful for emergencies, like server raids for example. while i can very much see the appeal in making it opt-in by default, there's occasionally a certain role that only an `@everyone` ping can fulfill. making these opt-in can occasionally just be adding additional hoops for people to jump through, which could be incredibly annoying.

i think it'd make more sense to have `@everyone` opt-out by default, *but only in small servers* (<50 people). perhaps this could also apply for channels where only a small number of people have read access to.

[the remark this answers is not in the document]
Comment (NinekoTheCat) [reply rfcs/rfc-0004-role-and-everyone-mentions.md:83]: I like that idea

**Comment by NinekoTheCat** [comment]
?

**Comment by NinekoTheCat** [comment]
well I said why the idea in my opinion isn't suited for development right now??

**Comment by NinekoTheCat** [comment]
I don't follow @nulldg

**Comment by NinekoTheCat** [comment]
@nulldg  I'd like to talk in dms about this so if you are in the Revolt Lounge send @Klepsi a friend request and we'll talk

**Comment by Rexogamer** [comment]
For the sake of other readers, can y'all refrain from randomly deleting your comments? It makes the conversation incredibly hard to follow.

File: rfcs/rfc-0004-role-and-everyone-mentions.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (NinekoTheCat) [reply rfcs/rfc-0004-role-and-everyone-mentions.md]: since I removed the mention of them I will talk about it here

basically I think that ping groups add a lot of entities that the frontend has to send to the backend such as a possible selector object? I don't quite know how it'd pan out and it's radically different enough to warrant a separate RFC

**Comment by JackDotJS** [comment]
i wonder if it'd also make sense to have some kind of rate limit on mass pings. not only could this help reduce server load, but it could also reduce spam and annoyance in general. ideally, it'd be something server owners could control, similar to slowmode on discord.

**Comment by NinekoTheCat** [comment]
> i wonder if it'd also make sense to have some kind of rate limit on mass pings. not only could this help reduce server load, but it could also reduce spam and annoyance in general. ideally, it'd be something server owners could control, similar to slowmode on discord.

so you're proposing like a configurable per role limit on mentions?

**Comment by JackDotJS** [comment]
i was thinking more of a server-wide rate limit, but per-role and maybe even per-channel would be nice for more precise control.

**Comment by JackDotJS** [comment]
in fact, a rate limit on pings in general might be quite nice as an anti-spam measure... making it exclusively for role/everyone pings might just encourage spammers to use individual pings (which is also a problem over on Discord)

**Comment by NinekoTheCat** [comment]
this RFC only covers role and everyone pings so pleae concider it within that boundry, while I do agree a limit should be applied to pings, should it be set by server admins or be enabled by default in the first place?

**Comment by JackDotJS** [comment]
yeah the rate limit idea could probably use its own RFC.

in this context though, i'd probably have it enabled by default, but with a high limit. it'd provide *some* kind of anti-spam, but a very mild one (which would likely be enough for most people)

**Comment by NinekoTheCat** [comment]
like 10 mentions per minute?

**Comment by JackDotJS** [comment]
maybe more like 5
realistically, who needs more than that?

**Comment by NinekoTheCat** [comment]
5 it is I'll edit the rfc

**Comment by JackDotJS** [comment]
the only other concern i can think of right now is how mass pings would be handled on larger servers.
like, surely at some point you'd be fetching so many people from the database, you might just end up running out of memory, right? or at the very least, slowing the entire service down to a crawl.

**Comment by NinekoTheCat** [comment]
we can have pings be acknowliged the same way messages are

**Comment by xXBuilderBXx** [comment]
> the only other concern i can think of right now is how mass pings would be handled on larger servers. like, surely at some point you'd be fetching so many people from the database, you might just end up running out of memory, right? or at the very least, slowing the entire service down to a crawl.

Not really because you only need the ID of the user to actually ping (or selected role of user for role mentions) and not the entire data, also it could queue up the mentions.

**Comment by NinekoTheCat** [comment]
> > the only other concern i can think of right now is how mass pings would be handled on larger servers. like, surely at some point you'd be fetching so many people from the database, you might just end up running out of memory, right? or at the very least, slowing the entire service down to a crawl.
> 
> Not really because you only need the ID of the user to actually ping (or selected role of user for role mentions) and not the entire data, also it could queue up the mentions.

defer means queue, that's what I suggested I think

**Comment by rethyria** [comment]
Users should always have full control of how they are being contacted and notified. This is where I think Discord fails on this topic.

While default notification settings for a server can be set by a server admin, there is no default option for a user. This means the user needs to adjust settings for every server they join - which can be hundreds.

Default notification settings should be configurable by the user so that they can adjust notifications for every server by default. Settings could then be configured per server or channel on a case-by-case basis by the user. 

A few other points:
- If role/everyone pings will be disabled automatically once certain conditions are met, the server admin should be notified when this happens
- Should there be any consideration for an '@'here ping?
- Should the role/everyone/here be case-sensitive or case-insensitive? Imo, the search should at least be case-insensitive
- Regarding this small server 50 people business, is it worth considering some sort of actual rules and terminology to classify server types (e.g. Personal and Community) and use this classification instead?
- There is currently no 'inbox' in Revolt so it is difficult to find mentions from some time ago, especially on active servers or if you were offline at the time of the ping
  - An inbox is a bit beyond the scope of this change but worth considering as a next step

**Comment by SpamixOfficial** [comment]
I think @EnigmaEmmy's points are good suggestions.

**Comment by Andre601** [comment]
Joining in late to give my two cents to this topic.
Generally, I'm fine with the idea of implemention everyone and role mentions, but I also would like to propose the following adjustments:

- Make role and everyone mentions off by default, no matter the server size. This is something I consider a generally bad habit of Discord, where the everyone mention is enabled by default... for the everyone role no less, meaning everyone can tag everyone, allowing chaos to be created from this.
  - In case of permissions, I would say that role mention and everyone mention perms would be made separately, as there can be cases where you want people to be able to tag roles, but not everyone. This is a major flaw in Discord's design of permissions imo that Revolt should avoid. Also, taggable roles should be tetermined by a role setting to enable or disable this. This would allow greater freedom of what roles one should be able to tag.
- The suggested syntax is confusing and overall just bad to have. It would be especially weird for people switching from Discord, who may be more familiar with their syntax, which is `<@&<id>>` for roles and `@everyone` for everyone mentions. A similar syntax should be used here. It wouldn't conflict with the ID system of Revolt as the ULID doesn't seem to use special characters such as ampersand, allowing easy detection of specific mention types. A `<*:@>` would just look weird in the client, especially given it doesn't render mentions at all.
- I suggest the addition of a `@Active` mention. This would be the same as `@Here` on discord, tagging people in a channel that are currently online. However, the name would it make it somewhat more clear of who it would tag: People that are "active".
  - Additionally can/should it be considered what users should be tagged, based on their status. A status like "focused" should perhaps be completely excluded from any mention to not disrupt their current work they may have.

As a final note would I also like to suggest considering the implementation of silent/supressed mentions similar to the `@silent` feature offered by Discord, to allow mentioning people but not giving them actual notifications.
Tho, I believe this should be its own RFC. If agreed, I can work out one to then discuss over.
